Added gulp setup command

timneutkens opened the pull request:
See #47 
Removed gulp symlink as it's not relevant anymore.

Igloczek commented on tasks/setup.js:
```diff
@@ -0,0 +1,51 @@
+  // Global variables
+  var plugins = this.opts.plugins,
+      config  = this.opts.configs,
+      fs      = require('fs-extra'),
```
We should add this in `gulpfile.js` and map `fs-extra` to `fs` and then use as a `plugins.fs` across the project.

BTW. It's werid, why they not merge this to node `js` package...

timneutkens replied:
Will do that tonight. Yeah it's quite a good library. They probably couldn't be bothered to do so. There are a lot of packages that provide the functionality fs-extra provides. Same for things like rimraf (which is actually included in fs-extra)

Igloczek replied:
Great, thank you 👍

Igloczek commented on package.json:
```diff
@@ -13,6 +13,7 @@
     "autoprefixer": "6.3.6",
     "browser-sync": "2.12.8",
     "cssnano": "3.5.2",
+    "fs-extra": "^0.30.0",
```
Let's use here a strict version `"0.30.0" - sometimes it leads to weird situations.

timneutkens replied:
Added the strict dependency on fs-extra. Might be good to enable https://greenkeeper.io on the package if we're doing this. Will create a seperate issue for that 😉

timneutkens commented:
Just noticed themes.json is required for the setup task to load. Will have a try to fix this..

timneutkens commented:
Been quite busy that last week, passed Magento front-end dev exam and such 😉 Will have a go at this tomorrow.

timneutkens commented:
So I did a few things.. I noticed `var` and `const` were used together. To be fully compatible with es2015 I changed all of them to `const` / `let`. To use `let` in node4 strict mode is required. So I also added that.

The other thing I did was remove the error thrown in the config loader because it causes any commands that don't use themes.json to crash when it's not available. Now only commands that actually use themes.json get the error. `gulp setup` wouldn't be available otherwise.

Igloczek commented on helpers/error-message.js:
```diff
@@ -0,0 +1,8 @@
+'use strict';
+module.exports = function (plugins) {
+  return function (message) {
+    return plugins.util.colors.red('\n' + '='.repeat(message.length) + ' \n')
```
This leads to weird looking output on smaller console windows. I often stick console as a "sidebar" and have about ~200-300px width - in this case I'd have few lines of `=` before msg.

timneutkens replied:
Added a max length of 50 that ok?

Igloczek commented on helpers/get-themes.js:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    return themeName ? themes[themeName] : themes;
+  }
```
I'd simplify this and just add `required` param to `config-loader`.
Passing theme name to function named `getThemes` is weird, b/c it should return themes object, not a single theme or error about this theme.

Igloczek replied:
Also in future we may have more configs required to run some tasks, so one solution to throw errors while something goes wrong, will be much better.

Igloczek commented on package.json:
```diff
@@ -14,7 +14,7 @@
     "browser-sync": "~2.13.0",
     "cssnano": "~3.7.3",
-    "fs-extra": "~0.30.0",
+    "fs-extra": "^0.30.0",
```
Change this back to `~`

Igloczek commented on gulpfile.js:
```diff
@@ -1,21 +1,25 @@
   'projectPath': plugins.fs.realpathSync('../../../') + '/'
 };
 
 config.themes = require('./helpers/config-loader')('themes.json', plugins, config);
```
This one should be probably removed right now.

timneutkens replied:
Yeah was just thinking the same.. lol 😉  tried to work around it but that just feels hacky. Will have a new go this evening.

timneutkens commented on gulpfile.js:
```diff
@@ -27,7 +31,7 @@ require('gulp-task-loader')({
+  const theme = config.themes[name];
   theme.locale.forEach(locale => {
     gulp.task(
       theme.lang + ':' + name + ':' + locale,
```
@Igloczek are these tasks used anywhere? Removing config.themes from the gulpfile.js will cause this to stop functioning

Igloczek replied:
Sure... Whole styles compilation works on top of this tasks.

timneutkens replied:
whaha, that'll be problematic :wink: Maybe it's not such a good idea to remove config.themes from the gulpfile then.

Igloczek replied:
Yup, but just a little bit 😄 
I'd remove `get-themes` and improve `config-loader`, then everything should works fine.

timneutkens replied:
Since config loader may not throw error's for themes.json at the time it's called we still need get-themes. I improved get themes to always support `--theme <theme name>`  It's got 2 error throw conditions but that's fine actually, since we still want to stop things like `gulp styles` that use themes.json when themes.json is not available. Also throwing an `plugins.util.PluginError` doesn't do well bringing the error while it's not used inside a gulp task. Which this resolves 👍  Let me know what your thoughts on this are. I am more than happy to improve it further, just can't see another way than returning an empty object when themes.json is not available ATM. The changes to config-loader will now allow it to fail on all other config loads.

Igloczek replied:
I think if we have a task to setup configuration, throwing errors while `themes.json` doesn't exist might not be so useful anymore.
To be honest I don't like idea of `get-themes`. IMO it's pointless to keep this as separate helper.

My goal is to keep things simple, but in this case simplifying tools for noobs (sorry, but it's super easy to copy / create configuration in right place) are adding complexity to whole project.

timneutkens replied:
> I think if we have a task to setup configuration, throwing errors while themes.json doesn't exist might not be so useful anymore.

I agree, only reason I added it is for when the config file can't be found. Otherwise things like `gulp styles` will run fine but won't do anything ofcourse

> To be honest I don't like idea of get-themes. IMO it's pointless to keep this as separate helper.

The theme reading logic is used about 5 times, that's why I seperated it.

> My goal is to keep things simple, but in this case simplifying tools for noobs (sorry, but it's super easy to copy / create configuration in right place) are adding complexity to whole project.

I agree that we should keep things simple. The changes I made aren't so complex, it's a change in function indeed. But the only real change is that `config-loader` now has an optional parameter to disregard the file not being there. I also added a themes helper which replaces a function that is used 5 times and improved it with a check to make sure the provided theme in `--theme <theme name>` is defined.

Igloczek replied:
Ok, let's fix few minor issues mentioned in other comments and I'm going to merge this.

timneutkens replied:
Cool 👍 Will make those right away!

Igloczek commented on helpers/config-loader.js:
```diff
@@ -1,4 +1,9 @@
+module.exports = function (file, plugins, config, failOnError) {
+  if (typeof failOnError === 'undefined') {
+    failOnError = true;
+  }
```
You can use ES6 default parameter value instead this.

https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Functions/Default_parameters

timneutkens replied:
http://node.green/ we support node.js 4.x atm. It doesn't support this.

Igloczek replied:
😿

Igloczek commented on gulpfile.js:
```diff
@@ -1,21 +1,25 @@
 
-config.themes = require('./helpers/config-loader')('themes.json', plugins, config);
+config.themes = require('./helpers/config-loader')('themes.json', plugins, config, false);
+plugins.getThemes = require('./helpers/get-themes')(plugins, config);
```
By `plugins` I mean 3rd party dependencies, not helpers.

timneutkens replied:
The reason I added it this way is to expose the helper with less parameters. Also plugins is already shared across everything.

Igloczek commented on helpers/error-message.js:
```diff
@@ -0,0 +1,9 @@
+      + plugins.util.colors.yellow(message)
+      + plugins.util.colors.red('\n' + '='.repeat(lineLength))
+  }
+};
```
New line on file end.

timneutkens replied:
Done.

Igloczek commented on helpers/get-themes.js:
```diff
@@ -0,0 +1,24 @@
+
+    return themeName ? themes[themeName] : themes;
+  }
+};
```
New line on file end.

timneutkens replied:
Done.

Igloczek commented on helpers/less.js:
```diff
@@ -1,14 +1,15 @@
+          maps       = plugins.util.env.maps || false,
+          production = plugins.util.env.prod || false,
+          lessFiles  = file || [],
+          postcss    = [],
```
`postcess` should be editable

timneutkens replied:
See https://github.com/SnowdogApps/magento2-frontools/pull/54#discussion_r72061542

Igloczek commented on helpers/scss.js:
```diff
@@ -1,13 +1,14 @@
+          dest       = config.projectPath + theme.dest + '/' + locale,
+          maps       = plugins.util.env.maps || false,
+          production = plugins.util.env.prod || false,
+          postcss    = [],
```
`postcss` should be editable.

timneutkens replied:
`const` prevents a variable from being reassign like so `postcss = ['example']` `postcss.push('example')` is perfectly fine 👍 same goes for objects. Assigning keys is allowed. Just not reassigning. the variable. That's why you can use const almost everywhere in the codebase. Just a few exceptions. Mathias Bynens has a great article about it: https://mathiasbynens.be/notes/es6-const

Igloczek replied:
Thanks for clarifying, I'm new in ES6 and thought const is const, not almost const 😄

timneutkens replied:
Cool, coming from PHP I thought the same 😄
